command/genesis/polybft_params.go

```diff
@@ -0,0 +1,224 @@
+	smartContracts, err := deployContracts()
+	if err != nil {
+		return nil, err
+	}
+
+	config := &polybft.PolyBFTConfig{
+		// TODO: Bridge
+		InitialValidatorSet: validators,
+		BlockTime:           p.blockTime,
+		EpochSize:           p.epochSize,
+		SprintSize:          p.sprintSize,
+		ValidatorSetSize:    p.validatorSetSize,
+		ValidatorSetAddr:    types.Address(ValidatorSetAddr),
+		SidechainBridgeAddr: types.Address(SidechainBridgeAddr),
+		SmartContracts:      smartContracts,
+	}
+
+	return config, nil
+}
+
+func (p *genesisParams) GetChainConfig() (*chain.Chain, error) {
+	validatorsInfo, err := ReadValidatorsByRegexp(path.Dir(p.genesisPath), p.validatorPrefixPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Predeploy staking smart contracts
+	genesisValidators := p.getGenesisValidators(validatorsInfo)
+
+	polyBftConfig, err := p.getPolyBftConfig(genesisValidators)
+	if err != nil {
+		return nil, err
+	}
+
+	extra := polybft.Extra{Validators: GetInitialValidatorsDelta(validatorsInfo)}
+
+	chainConfig := &chain.Chain{
+		Name: p.name,
+		Genesis: &chain.Genesis{
+			GasLimit:   p.blockGasLimit,
+			Difficulty: 0,
+			Alloc:      map[types.Address]*chain.GenesisAccount{},
+			ExtraData:  append(make([]byte, 32), extra.MarshalRLPTo(nil)...),
+			GasUsed:    command.DefaultGenesisGasUsed,
+			Mixhash:    polybft.PolyMixDigest,
+		},
+		Params: &chain.Params{
+			ChainID: int(p.chainID),
+			Forks:   chain.AllForksEnabled,
+			Engine: map[string]interface{}{
+				string(server.PolyBFTConsensus): polyBftConfig,
+			},
+		},
+		Bootnodes: p.bootnodes,
+	}
+
+	// set generic validators as bootnodes if needed
+	if len(p.bootnodes) == 0 {
+		for i, validator := range validatorsInfo {
+			// /ip4/127.0.0.1/tcp/10001/p2p/16Uiu2HAm9r5oP8Dmfsqbp1w2LdPU4YSFggKvwEmT6aTpWU8c8R13
+			bnode := fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", "127.0.0.1", bootnodePortStart+i, validator.NodeID)
+			chainConfig.Bootnodes = append(chainConfig.Bootnodes, bnode)
+		}
+	}
+
+	// Premine accounts
+	if err := fillPremineMap(chainConfig.Genesis.Alloc, p.premine); err != nil {
+		return nil, err
+	}
+
+	return chainConfig, nil
+}
+
+func (p *genesisParams) getGenesisValidators(validators []GenesisTarget) (result []*polybft.Validator) {
+	if len(p.validators) > 0 {
+		for _, validator := range p.validators {
+			parts := strings.Split(validator, ":")
+			if len(parts) != 2 || len(parts[0]) != 32 || len(parts[1]) < 2 {
+				continue
+			}
+
+			result = append(result, &polybft.Validator{
+				Address: types.Address(ethgo.HexToAddress(parts[0])),
+				BlsKey:  parts[1],
+			})
+		}
+	} else {
+		for _, validator := range validators {
+			pubKeyMarshalled := validator.Account.Bls.PublicKey().Marshal()
+
+			result = append(result, &polybft.Validator{
+				Address: types.Address(validator.Account.Ecdsa.Address()),
+				BlsKey:  hex.EncodeToString(pubKeyMarshalled),
+			})
+		}
+	}
+
+	return result
+}
```

Comment:
The method as I see it is supposed to retrieve the validator set, but either from the validator set defined with `p.validators` or the passed in params. Why is this logic in the same method?
It's convoluted and shouldn't be bundled, please break it up

Reply:
Is it the suggestion to break the code into two separate functions? I would like to avoid splitting things into too many functions. This looks okay to me.